### ISSUE TITLE
execinfra: RemoteProducerMetadata_Metrics race

### ISSUE THREAD

**Issue by asubiotto** [cockroachdb/cockroach#52890]
In #52463, I encountered an interesting race in CI ([build link](https://teamcity.cockroachdb.com/viewLog.html?buildId=2182951&buildTypeId=Cockroach_UnitTests)):
```
------- Stdout: -------
=== RUN   TestUncertaintyErrorIsReturned/vectorize=true/SELECT_*_FROM_t_JOIN_onerow_ON_t.x_=_onerow.x
==================
WARNING: DATA RACE
Write at 0x00c0076c29d0 by goroutine 962:
  github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*ColBatchScan).DrainMeta()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/colbatch_scan.go:114 +0x38a
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340 +0xdb
  github.com/cockroachdb/cockroach/pkg/sql/colexec.(*HashRouter).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:727 +0x10d
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupRouter.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:663 +0xf2
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1189 +0x51

Previous write at 0x00c0076c29d0 by goroutine 666:
  github.com/cockroachdb/cockroach/pkg/sql/colfetcher.(*ColBatchScan).DrainMeta()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colfetcher/colbatch_scan.go:114 +0x38a
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:340 +0xdb
  github.com/cockroachdb/cockroach/pkg/sql/colexec.(*HashRouter).Run()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/routers.go:727 +0x10d
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupRouter.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:663 +0xf2
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1.1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1189 +0x51

Goroutine 962 (running) created at:
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1188 +0x86
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.StartableFn.Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:47 +0x5b
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:321 +0x330
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:346 +0x98
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Start()
      <autogenerated>:1 +0x7f
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).runFlowNow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:109 +0x26a
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:130 +0x431
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:326 +0x150
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:127 +0xcb
  github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:574 +0x4bc
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:630 +0xa1
  github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:44 +0xbd8
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:211 +0x135
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:61 +0xd9
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:55 +0x80
  google.golang.org/grpc.chainUnaryServerInterceptors.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:907 +0x110
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:632 +0x1d8
  google.golang.org/grpc.(*Server).processUnaryRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1082 +0x954
  google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1405 +0x12bd
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xc8

Goroutine 666 (finished) created at:
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1188 +0x86
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.StartableFn.Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:47 +0x5b
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).startInternal()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:321 +0x330
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Start()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:346 +0x98
  github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Start()
      <autogenerated>:1 +0x7f
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).runFlowNow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:109 +0x26a
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:130 +0x431
  github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr()
      /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:326 +0x150
  github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowScheduler).ScheduleFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow_scheduler.go:127 +0xcb
  github.com/cockroachdb/cockroach/pkg/sql/distsql.(*ServerImpl).SetupFlow()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql/server.go:574 +0x4bc
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:630 +0xa1
  github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc/server.go:44 +0xbd8
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:211 +0x135
  google.golang.org/grpc.getChainUnaryHandler.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:921 +0x146
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:61 +0xd9
  github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor-fm()
      /go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:55 +0x80
  google.golang.org/grpc.chainUnaryServerInterceptors.func1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:907 +0x110
  github.com/cockroachdb/cockroach/pkg/sql/execinfrapb._DistSQL_SetupFlow_Handler()
      /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/api.pb.go:632 +0x1d8
  google.golang.org/grpc.(*Server).processUnaryRPC()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1082 +0x954
  google.golang.org/grpc.(*Server).handleStream()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1405 +0x12bd
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:746 +0xc8
==================
FAIL	github.com/cockroachdb/cockroach/pkg/sql/rowexec	19.818s
```

I don't believe the PR caused this race. The race itself is two writes to the same object here:
https://github.com/cockroachdb/cockroach/blob/8324ed363b89ea2bd301f0245dd9a4f621a55c2f/pkg/sql/colfetcher/colbatch_scan.go#L114

The reason I find this interesting is that the object is requested in the line right above the write. `sync.Pool` is safe for concurrent use and should return an object safe for use by the requesting goroutine. @nvanbenschoten, could this be similar to #46793?

**Comment by nvanbenschoten** [comment]
> @nvanbenschoten, could this be similar to #46793?

It could be, but I wouldn't jump to that immediately. We've disabled the use of finalizers in Pebble, which is why #46793 hasn't flaked since June. So I'd assume it's not until we get a bit more information.

Have you explored whether anyone is double-freeing into the pool? That could explain something like this.

**Comment by asubiotto** [comment]
Looks like it most likely is not that, the test is failing pretty frequently on master after I merged that PR (#52948), so I'll have to investigate further. I don't see any double-frees and wasn't able to reproduce, but I'll dig a little deeper.

**Comment by asubiotto** [comment]
It was indeed a double-free, thanks for the pointer.
